assembly/src/assembler/module_graph/mod.rs

```diff
@@ -408,17 +558,23 @@ impl ModuleGraph {
         self.topo.as_slice()
     }
 
-    /// Compute the topological sort of the callgraph rooted at `caller`
-    pub fn topological_sort_from_root(
+    /// Compute the topological sort of the callgraph rooted at `caller`, only for procedures that
+    /// need to be assembled (i.e. in `Ast` representation).
+    pub fn topological_sort_ast_procs_from_root(
```

Comment:
I wouldn't deal with this detail here in the `ModuleGraph` - just construct the topological sort as normal, and in the `Assembler`, handle each `GlobalProcedureIndex` based on what it resolves to:

* If it resolves to an already-compiled module, then make the decision on whether to incorporate that MAST into the MAST being built, or simply skip that procedure if it is being included by reference
* If it doesn't resolve to anything we have on hand, it must be a phantom procedure, so treat it as if it is included by reference
* If it resolves to a procedure in AST form, proceed with lowering it to MAST

Reply:
Moved the filtering for AST form outside the function - not 100% if you had something more involved in mind though.